backend/monoprocess/monoprocess.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/cschleiden/go-workflows/workflow"
+)
+
+type monoprocessBackend struct {
+	backend.Backend
+
+	workflowSignal chan struct{}
+	activitySignal chan struct{}
+	signalTimeout  time.Duration
+
+	logger *slog.Logger
+}
+
+// NewMonoprocessBackend wraps an existing backend and improves its responsiveness
+// in case the backend and worker are running in the same process. This backend
+// uses channels to notify the worker every time there is a new task ready to be
+// worked on. Note that only one worker will be notified.
+// IMPORTANT: Only use this backend when the backend and worker are running in
+// the same process.
+func NewMonoprocessBackend(b backend.Backend) *monoprocessBackend {
+	mb := &monoprocessBackend{
+		Backend:        b,
+		workflowSignal: make(chan struct{}, 1),
+		activitySignal: make(chan struct{}, 1),
+		logger:         b.Logger(),
+	}
+	return mb
+}
+
+func (b *monoprocessBackend) GetWorkflowTask(ctx context.Context) (*task.Workflow, error) {
+	// loop until either we find a task or the context is cancelled
+	for {
+		if w, err := b.Backend.GetWorkflowTask(ctx); w != nil || err != nil {
+			return w, err
+		}
+		b.logger.DebugContext(ctx, "worker waiting for workflow task signal")
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-b.workflowSignal:
+			b.logger.DebugContext(ctx, "worker got a workflow task signal")
+		}
+	}
+}
+
+func (b *monoprocessBackend) GetActivityTask(ctx context.Context) (*task.Activity, error) {
+	// loop until either we find a task or the context is cancelled
+	for {
+		if a, err := b.Backend.GetActivityTask(ctx); a != nil || err != nil {
+			return a, err
+		}
+		b.logger.DebugContext(ctx, "worker waiting for activity task signal")
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-b.activitySignal:
+			b.logger.DebugContext(ctx, "worker got an activity task signal")
+		}
+	}
+}
+
+func (b *monoprocessBackend) CreateWorkflowInstance(ctx context.Context, instance *workflow.Instance, event *history.Event) error {
+	if err := b.Backend.CreateWorkflowInstance(ctx, instance, event); err != nil {
+		return err
+	}
+	b.notifyWorkflowWorker(ctx)
+	return nil
+}
+
+func (b *monoprocessBackend) CompleteWorkflowTask(
+	ctx context.Context,
+	task *task.Workflow,
+	instance *workflow.Instance,
+	state core.WorkflowInstanceState,
+	executedEvents, activityEvents, timerEvents []*history.Event,
+	workflowEvents []history.WorkflowEvent,
+) error {
+	if err := b.Backend.CompleteWorkflowTask(ctx, task, instance, state, executedEvents, activityEvents, timerEvents, workflowEvents); err != nil {
+		return err
+	}
+
+	hasWorkflowTasks := false
+	hasActivityTasks := false
+
+	for _, e := range executedEvents {
+		if e.Type == history.EventType_WorkflowTaskStarted {
+			hasWorkflowTasks = true
+			break
+		}
+	}
+	for _, e := range activityEvents {
+		if e.Type == history.EventType_ActivityScheduled {
+			hasActivityTasks = true
+			break
+		}
+	}
```

Comment:
just checking `len(activityEvents) > 0` should be enough, otherwise this makes sense